maup/repair.py

```diff
@@ -1,27 +1,1217 @@
+        reconstructed_index_by_iloc = dict(
+            (i, list(reconstructed_df.index)[i]) for i in range(len(reconstructed_df))
+        )  # Build indexing dictionary
+
+        for keeper in holes_to_keep:
+            possible_geom_integer_indices = [
+                *set(numpy.ndarray.flatten(reconstructed_spatial_index.query(keeper)))
+            ]
+            possible_geom_indices = [
+                reconstructed_index_by_iloc[k] for k in possible_geom_integer_indices
+            ]
+            for g_ind in possible_geom_indices:
+                reconstructed_df["geometry"][g_ind] = reconstructed_df["geometry"][
+                    g_ind
+                ].difference(keeper)
+
+    if min_rook_length is not None:
+        # Find all inter-polygon boundaries shorter than min_rook_length and replace them
+        # with queen adjacencies by manipulating coordinates of all surrounding polygon.
+        print("Converting small rook adjacencies to queen...")
+        reconstructed_df = small_rook_to_queen(reconstructed_df, min_rook_length)
+
+    if isinstance(geometries0_df, geopandas.GeoDataFrame):
+        return reconstructed_df
+    else:
+        return reconstructed_df.geometry
+
+
+# SUPPORTING FUNCTIONS:
+
+# Partition the region via ALL precinct boundaries, identify each polygon in the
+# partition according to which polygons in the original intersected to create it,
+# and organize this data according to order of the overlaps.  (Order zero = hole)
+
+
+def building_blocks(geometries0_df):
+    if isinstance(geometries0_df, geopandas.GeoDataFrame) == False:
+        raise TypeError(f"Input must be a GeoDataFrame!")
+
+    geometries_df = geometries0_df.copy()
+
+    # Make a list of all the boundaries of all the polygons.
+    # This won't work properly with multi-polygons, so explode first:
+
+    boundaries = []
+    geometries_exploded_df = geometries_df.explode(index_parts=False).reset_index(
+        drop=True
+    )
+    for i in geometries_exploded_df.index:
+
+        boundaries.append(shapely.boundary(geometries_exploded_df["geometry"][i]))
+
+    boundaries_union = unary_union(boundaries)
+
+    # Create geodataframe with all the pieces created by overlaps of all orders,
+    # together with a set for each piece consisting of the polygons that created the overlap.
+
+    pieces_df = geopandas.GeoDataFrame(
+        columns=["polygon indices"],
+        geometry=geopandas.GeoSeries([geom for geom in polygonize(boundaries_union)]),
+        crs=geometries_df.crs,
+    )
+
+    for i in pieces_df.index:
+        pieces_df["polygon indices"][i] = set()
+
+    g_spatial_index = STRtree(
+        geometries_df["geometry"]
+    )  # Build STRtree for the main geometries
+    g_index_by_iloc = dict(
+        (i, list(geometries_df.index)[i]) for i in range(len(geometries_df))
+    )  # Build indexing dictionary
+
+    print("Identifying overlaps...")
+    for i in progress(pieces_df.index, len(pieces_df.index)):
```

Comment:
The presence of a progress bar suggests that this is The Slow Part.  Before we release this, we should benchmark it against the old version and see if there are any significant performance regressions. I don't think a minor slowdown should stop us from releasing—we can always fine-tune performance later!—but I'd like to make sure it's not _dramatically_ worse due to the relative complexity (lines of code complexity, not big-O complexity ;)) of this new algorithm.